keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,137 @@
+
+### Pod Failure Cases
+
+#### RestartPolicy of Pods as Never
+
+Whenever a Pod of a Job fails, the Job will recreate a Pod as replacement against the failed Pod. The Job will continue this process of creating new Pods until the Job reaches its `backoffLimit`(by default the backoffLimit is `6`). The newly created Pod against the failed Pod will reuse the resources of the failed Pod. Once the Job reaches its `backoffLimit` and does not have the required successful `.spec.completions` count, the Job is termed as a `Failed` Job. When the Job is marked as failed, no more Pods would be created by the Job. So, the remaining owned resources of the Job will be reclaimed by queue. 
+
+#### RestartPolicy of Pods as OnFailure
+
+When the Pods of the Job have `.spec.template.spec.restartPolicy = "OnFailure"`, the failed Pod stays on the node and the failed container within the Pod is restarted. The Pod might run indefinitely or get marked as `Failed` after the `.spec.activeDeadlineSeconds` is completed by Job if specified. In this case also, Kueue won't reclaim the resources of the Job until the Job gets completed as either `Failed` or `Succeeded`.
+
+Hence, as seen from the above discussed Pod failure cases, we conclude that the Pods of a Job which fail during its execution, its resources are not immediately reclaimed by Kueue. Only when the Job gets marks as `Failed`, the resources of failed Pods will be reclaimed by Kueue.
+
+## API
+
+A new field `ReclaimedPodSets` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+    // conditions hold the latest available observations of the Workload
+    // current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    Conditions []WorkloadCondition `json:"conditions,omitempty"`
+
+    // list of count of Pods of a PodSet with resources reclaimed
+    // +optional
+    ReclaimedPodSets []ReclaimedPodSetPods `json:"reclaimedPodSets"`
+}
+
+// ReclaimedPodSetPod defines the PodSet name and count of successfully completed Pods 
+// belonging to the PodSet whose resources can be reclaimed.
+type ReclaimedPodSet struct{
+    Name string
+    Count int
+}
+```
+
+## Implementation
+
+Kueue's job reconciler will compare the Job's `.status.Succeeded` field and the Workload's `.status.reclaimedPodSets[i].count` field value. If the former value is greater than the later then, Kueue's Job reconciler will update the Workload object's `.status`. Workload reconciler will catch the update event of Kueue Job reconciler and release the resources of the newly succeeded Pods to the ClusterQueue depending upon the case the Job satisfies discussed in the section of [successful Pod completion](#pod-successful-completion-cases).
```

Comment:
we don't need to repeat the formula, but the first sentence of this paragraph seems to contradict what you said in the analysis above.

Or maybe I'm not understanding your proposal. I think all the calculations will have to be in the job controller. The workload controller already has the curated number of reclaimable pods to update the ClusterQueues usage.

Reply:
> we don't need to repeat the formula, but the first sentence of this paragraph seems to contradict what you said in the analysis above.

In the analysis above, comparison of values of Job's `.spec.parallelism` and it's remaining completion count (**M**) is explained, and at what condition in future the resources will reclaimed by Kueue. Here, in the first sentence, I'm trying to say that, if the mentioned values differ, then workload object will be updated. And the actual cases where the decision to reclaim or not will be taken by workload reconciler when it recieves an update event from job reconciler.

> I think all the calculations will have to be in the job controller. The workload controller already has the curated number of reclaimable pods to update the ClusterQueues usage.

What I have assumed is, during the handling of update event of workload by its reconciler, old and new workloads objects are available. 

1. Calculate the difference in the count of `.status.podSetReclaims` of  old and new workload object.
2. Check the job parallelism and remaining completion values (Here a GET call to fetch Job will be required).
3. Calculate the effective count of pods to reclaim and update cache.

I see the issue with my implementation approach here. There could be inconsistency in Job's remaining completion value.

What you have said on having all calculations in job controller seems the correct to me now. But, I see one challenge while releasing the resources from workload reconciler. How should the Job reconciler convey the effective count of Pods to be reclaimed to workload reconciler. As the workload reconciler has a cache object and the Job reconciler does not.
Can having cache object as a part of job reconciler correct in this case?

Reply:
The job controller doesn't need access to the cache (and it shouldn't have).

The job controller can simply write into the workload status how many pods can be reclaimed.

The workload controller observes that change and updates the cache.